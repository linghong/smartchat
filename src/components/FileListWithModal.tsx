import { useState, FC } from 'react';
import { X } from 'lucide-react';

import FileModal from '@/src/components/FileModal';
import FileThumbnailSelector from '@/src/components/FileThumbnailSelector';

import { FileData } from '@/src/types/chat';

interface FileListWithModalProps {
  fileSrc: FileData[] | undefined;
  handleFileDelete: (index: number) => void;
  isDeleteIconShow?: boolean;
}

const FileListWithModal: FC<FileListWithModalProps> = ({
  fileSrc,
  handleFileDelete,
  isDeleteIconShow = false
}) => {
  const [modalOpen, setModalOpen] = useState<boolean>(false);
  const [selectedFile, setSelectedFile] = useState<FileData | undefined>();

  const openModal = (fileSrc: FileData) => {
    setSelectedFile(fileSrc);
    setModalOpen(true);
  };

  const closeModal = () => {
    setModalOpen(false);
    setSelectedFile(undefined);
  };

  if (!fileSrc || fileSrc.length === 0) {
    return null;
  }

  return (
    <>
      <div className="flex flex-row justify-start w-full p-2">
        {fileSrc.map((fileData: FileData, i: number) => (
          <div key={i} className="w-20 h-full mx-2 relative">
            <div
              className="mr-2 p-1 bg-clip-border bg-slate-100 border-2 border-violet-100"
              onClick={() => openModal(fileData)}
              onKeyDown={e => e.key === 'Enter' && openModal(fileData)}
              aria-label={`Click to view larger file ${i + 1}`}
              title={`Thumbnail for uploaded file ${i + 1}`}
              tabIndex={0}
              role="button"
            >
              <FileThumbnailSelector fileData={fileData} />
            </div>
            {isDeleteIconShow && (
              <button
                onClick={() => handleFileDelete(i)}
                className="absolute -top-2 -right-2 bg-white rounded-full p-1 shadow-md"
                aria-label={`Delete file ${i + 1}`}
                role="button"
              >
                <X className="text-gray-600" />
              </button>
            )}
          </div>
        ))}
      </div>
      {modalOpen && selectedFile && (
        <FileModal fileData={selectedFile} onClose={closeModal} />
      )}
    </>
  );
};

export default FileListWithModal;
