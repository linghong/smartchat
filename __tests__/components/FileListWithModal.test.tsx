import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';

import FileListWithModal from '@/src/components/FileListWithModal';

describe('FileListWithModal', () => {
  const fileSrc = [
    {
      base64Content: 'data:image/png;base64,base64Imagedata1',
      type: 'image/png',
      size: 5000,
      name: 'image1'
    },
    {
      base64Content: 'data:image/png;base64,base64Imagedata2',
      type: 'image/png',
      size: 8000,
      name: 'image2'
    }
  ];

  const handleFileDelete = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render the list of images', () => {
    render(
      <FileListWithModal
        fileSrc={fileSrc}
        handleFileDelete={handleFileDelete}
      />
    );

    const files = screen.getAllByRole('button', {
      name: /click to view larger file/i
    });
    expect(files).toHaveLength(fileSrc.length);
  });

  it('should open the modal when an image is clicked', () => {
    render(
      <FileListWithModal
        fileSrc={fileSrc}
        handleFileDelete={handleFileDelete}
      />
    );

    const image = screen.getByRole('button', {
      name: /click to view larger file 1/i
    });
    fireEvent.click(image);

    const modal = screen.getByRole('dialog');
    expect(modal).toBeInTheDocument();

    const modalImage = screen.getByAltText('Expanded view');
    expect(modalImage).toBeInTheDocument();
    expect(modalImage).toHaveAttribute('src', fileSrc[0].base64Image);
  });

  it('should open the modal when Enter key is pressed', () => {
    render(
      <FileListWithModal
        fileSrc={fileSrc}
        handleFileDelete={handleFileDelete}
      />
    );

    const image = screen.getByRole('button', {
      name: /click to view larger file 1/i
    });
    fireEvent.keyDown(image, { key: 'Enter', code: 'Enter' });

    const modal = screen.getByRole('dialog');
    expect(modal).toBeInTheDocument();

    const modalImage = screen.getByAltText('Expanded view');
    expect(modalImage).toBeInTheDocument();
    expect(modalImage).toHaveAttribute('src', fileSrc[0].base64Image);
  });

  it('should close the modal when the close button is clicked', () => {
    render(
      <FileListWithModal
        fileSrc={fileSrc}
        handleFileDelete={handleFileDelete}
      />
    );

    const image = screen.getByRole('button', {
      name: /click to view larger file 1/i
    });
    fireEvent.click(image);

    const closeButton = screen.getByRole('button', {
      name: /close image view/i
    });
    fireEvent.click(closeButton);

    expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
  });

  it('should display the delete icon when isDeleteIconShow is true', () => {
    render(
      <FileListWithModal
        fileSrc={fileSrc}
        handleFileDelete={handleFileDelete}
        isDeleteIconShow={true}
      />
    );

    const deleteIcons = screen.getAllByLabelText(/Delete file/i);
    expect(deleteIcons).toHaveLength(fileSrc.length);
  });

  it('should not display the delete icon when isDeleteIconShow is false', () => {
    render(
      <FileListWithModal
        fileSrc={fileSrc}
        handleFileDelete={handleFileDelete}
        isDeleteIconShow={false}
      />
    );

    const deleteIcons = screen.queryAllByLabelText(/Delete file/i);
    expect(deleteIcons).toHaveLength(0);
  });

  it('should call handleFileDelete when the delete icon is clicked', () => {
    render(
      <FileListWithModal
        fileSrc={fileSrc}
        handleFileDelete={handleFileDelete}
        isDeleteIconShow={true}
      />
    );

    const deleteIcon = screen.getByLabelText('Delete file 1');
    fireEvent.click(deleteIcon);

    expect(handleFileDelete).toHaveBeenCalledTimes(1);
    expect(handleFileDelete).toHaveBeenCalledWith(0);
  });

  it('should match snapshot', () => {
    const { asFragment } = render(
      <FileListWithModal
        fileSrc={fileSrc}
        handleFileDelete={handleFileDelete}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
