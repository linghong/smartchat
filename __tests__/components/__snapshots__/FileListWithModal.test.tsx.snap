// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileListWithModal should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-row justify-start w-full p-2"
  >
    <div
      class="w-20 h-full mx-2 relative"
    >
      <div
        aria-label="Click to view larger file 1"
        class="mr-2 p-1 bg-clip-border bg-slate-100 border-2 border-violet-100"
        role="button"
        tabindex="0"
        title="Thumbnail for uploaded file 1"
      >
        <div
          class="bg-no-repeat bg-center bg-contain min-h-[60px] h-full w-full cursor-pointer"
          style="background-image: url(data:image/png;base64,base64Imagedata1);"
        />
      </div>
    </div>
    <div
      class="w-20 h-full mx-2 relative"
    >
      <div
        aria-label="Click to view larger file 2"
        class="mr-2 p-1 bg-clip-border bg-slate-100 border-2 border-violet-100"
        role="button"
        tabindex="0"
        title="Thumbnail for uploaded file 2"
      >
        <div
          class="bg-no-repeat bg-center bg-contain min-h-[60px] h-full w-full cursor-pointer"
          style="background-image: url(data:image/png;base64,base64Imagedata2);"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
